### DISCUSSION TITLE
How to record multiple responses for a stimulus via "image-keyboard-response."

### DISCUSSION THREAD

**Discussion by K104Kaz** [jspsych/jsPsych#1302] [Q&A] [answered]
Hi, experts. 
I'm wandering that how can I get data for several response for a single image by using of "image-keyboard-response" plugin. When I set "response_ends_trial" parameter "false", the image remained but only the first response was recorded. 
I think there is no parameter for this in the plugin, so perhaps I should modify the plugin file? I'm new to jsPsych and javascript. Any comments would be appreciated !

**Comment by becky-gilbert** [comment, marked as answer]
Hi @K104Kaz, yes I think you'll want to modify the plugin file to do this:
- In the `jsPsych.pluginAPI.getKeyboardResponse` function, change `persist` from `false` to `true` (`false` means that the keyboard listener should be cancelled after the first response). 
```javascript
var keyboardListener = jsPsych.pluginAPI.getKeyboardResponse({
  callback_function: after_response,
  valid_responses: trial.choices,
  rt_method: 'performance',
  persist: true,   // <- changed from false to true
  allow_held_key: false
});
```
- Change the `responses` object so that it stores _arrays_ of rts and keys, instead of just one value for each:
```javascript
// store response
var response = {
  rt: [], // <- changed from null to empty arrays, so that you can store multiple values
  key: []
};
```
- In the `after_response` function, remove the `if` statement that checks if `response.key` is null, and if so, stores the response. Instead, you want to add `info.rt` and `info.key` onto the `responses.rt` and `responses.key` arrays after every response:
```javascript
response.rt.push(info.rt);   // <- these two lines instead of "if (response.key == null) {response = info;}"
response.key.push(info.key);
```
- In the `end_trial` function, when gathering the data in the `trial_data` object, use `JSON.stringify` to convert the rt and key arrays into JSON-formatted strings:
```javascript
var trial_data = {
  "rt": JSON.stringify(response.rt), // <- add JSON.stringify
  "stimulus": trial.stimulus,
  "key_press": JSON.stringify(response.key) // <- add JSON.stringify
}; 
```
I hope this helps! Feel free to ask questions here and post your code if you get stuck.

**Reply by K104Kaz** [reply]
Thank you! It worked without a problem!

**Reply by rMinion** [reply]
@becky-gilbert In the example above Is it possible to get the value of the response key before ending the trial?

**Reply by becky-gilbert** [reply]
Hi @rMinion, I'm not entirely sure what you mean... It is possible to get the response key during the trial, after each response, from inside the plugin. Specifically, the `callback_function` (`after_response` in the example above) is called immediately after every key press - it receives an object containing the value of the key that was pressed and the RT. So you can use that information however you need to inside the `after_response` function. 
But if you mean that you want to get the value of the key from outside of the plugin code (i.e. in your main experiment script) before the trial ends, I don't think that's easily done.

**Comment by xinyiguan** [comment]
I wonder if I can do the same (record multiple keyboard responses) for 1 stimuli via "audio-keyboard-responses."? I was trying to modified the audio-keyboard-responses plugin codes in a similar fashion. But I was not able to fully keep track and understand a few things. Specifically, I don't quite understand the following: 

In the `plugin-audio-keyboard-response.js` script starting from line 165:

            const setup_keyboard_listener = () => {
                // start the response listener
                if (context !== null) {
                    this.jsPsych.pluginAPI.getKeyboardResponse({
                        callback_function: after_response,
                        valid_responses: trial.choices,
                        rt_method: "audio",
                        persist: false,
                        allow_held_key: false,
                        audio_context: context,
                        audio_context_start_time: startTime,
                    });
                }
                else {
                    this.jsPsych.pluginAPI.getKeyboardResponse({
                        callback_function: after_response,
                        valid_responses: trial.choices,
                        rt_method: "performance",
                        persist: false,
                        allow_held_key: false,
                    });
                }
            };

Can you elaborate a bit what `rt_method` means? What is the difference between "performance" and "audio"?

Would it be possible if you can walk me through the steps to modified the necessary part of the script to achieve my goal? Thanks! 

BTW, I am using the 7.2 version

**Reply by jodeleeuw** [reply]
Hi @xinyiguan 

> Can you elaborate a bit what `rt_method` means? What is the difference between "performance" and "audio"?

The docs for the `getKeyboardResponse` method describe this a bit: https://www.jspsych.org/7.2/reference/jspsych-pluginAPI/#getkeyboardresponse. 

> Would it be possible if you can walk me through the steps to modified the necessary part of the script to achieve my goal? Thanks!

I think following the same steps described above by @becky-gilbert will work for the audio plugin too. The one change that you don't need to do the final one with `JSON.stringify`. Version 7.2 can handle arrays as the value for a data entry.

**Comment by phuongtran807** [comment]
Hi! I would like to make a similar alteration in my iat-html plugin. There is a function called 'force correct key response' when response.correct = false, so you have to press the right button in order to continue. In my case, I also want to record the second response time for pressing that button, so there are max. 2 values stored in the array.
I've tried it with the same instructions but there were several errors. For example did the blocks ended before every stimulus were presented and in general, the categorizations were off (i.e. the red x was presented even when pressing the right button). Also, in the final data all the trials were stored as correct (true).

Any help would be appreciated!

**Comment by matt-luke** [comment]
Much like @xinyiguan , I am trying to modify "audio-keyboard-responses" to accommodate multiple keypresses and their respective rt's. This time, in v8.1.0. I was wondering what I would have to do differently, due to the difference in the way the responses object is structured, as well as the after_response function. Thanks!
